src/infrastructure/acknowledgement.rs

```diff
@@ -0,0 +1,241 @@
+use crate::packet::OrderingGuarantee;
+use crate::sequence_buffer::SequenceBuffer;
+use crate::packet::SequenceNumber;
+use std::collections::HashMap;
+
+const REDUNDANT_PACKET_ACKS_SIZE: u16 = 32;
+
+/// Responsible for handling the acknowledgement of packets.
+pub struct AcknowledgementHandler {
+    // Local sequence number which we'll bump each time we send a new packet over the network
+    sequence_number: SequenceNumber,
+    // Using a Hashmap to track every packet we send out so we can ensure that we can resend when
+    // dropped.
+    sent_packets: HashMap<u16, SentPacket>,
+    // However, we can only reasonably ack up to REDUNDANT_PACKET_ACKS_SIZE + 1 packets on each
+    // message we send so this should be REDUNDANT_PACKET_ACKS_SIZE large.
+    received_packets: SequenceBuffer<ReceivedPacket>,
+
+    pub dropped_packets: Vec<SentPacket>,
```

Comment:
Since this is public we should have a comment here

Reply:
Maybe it would be a good idea to have a function `drain_dropped_packets()` which calls `drain` on the vector, just to have better encapsulation. Currently, you're calling this in `virtual_connection.rs` line 373 where you do the drain there.

Reply:
Love that idea 👍